File: instrumentation/core/src/main/java/de/mannodermaus/junit5/condition/DisabledOnManufacturerCondition.java
package de.mannodermaus.junit5.condition;

import android.annotation.TargetApi;
import android.os.Build;
import org.junit.jupiter.api.extension.ConditionEvaluationResult;
import org.junit.jupiter.api.extension.ExecutionCondition;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.platform.commons.util.Preconditions;

import java.util.Arrays;
import java.util.Optional;

import static de.mannodermaus.junit5.condition.EnabledOnManufacturerCondition.disabled;
import static de.mannodermaus.junit5.condition.EnabledOnManufacturerCondition.enabled;
import static org.junit.platform.commons.util.AnnotationUtils.findAnnotation;

class DisabledOnManufacturerCondition implements ExecutionCondition {

  private static final ConditionEvaluationResult ENABLED_BY_DEFAULT =
      ConditionEvaluationResult.enabled("@DisabledOnManufacturer is not present");

  @TargetApi(24)
  @Override
  public ConditionEvaluationResult evaluateExecutionCondition(ExtensionContext context) {
    Optional<DisabledOnManufacturer> optional = findAnnotation(context.getElement(), DisabledOnManufacturer.class);

    if (optional.isPresent()) {
      DisabledOnManufacturer annotation = optional.get();
      String[] patterns = annotation.value();
      boolean ignoreCase = annotation.ignoreCase();
      Preconditions.condition(patterns.length > 0, "You must declare at least one Manufacturer in @DisabledOnManufacturer");

      return Arrays.stream(patterns).anyMatch(value -> matchesCurrentManufacturer(value, ignoreCase))
          ? disabled()
          : enabled();
    }

    return ENABLED_BY_DEFAULT;
  }

  private boolean matchesCurrentManufacturer(String value, boolean ignoreCase) {
    if (ignoreCase) {
      return Build.MANUFACTURER.equalsIgnoreCase(value);
    } else {
      return Build.MANUFACTURER.equals(value);
    }
  }
}

File: instrumentation/core/src/main/java/de/mannodermaus/junit5/internal/DisabledOnManufacturerCondition.kt
package de.mannodermaus.junit5.internal

import androidx.annotation.RequiresApi
import android.os.Build
import de.mannodermaus.junit5.condition.DisabledOnManufacturer
import de.mannodermaus.junit5.internal.EnabledOnManufacturerCondition.Companion.disabled
import de.mannodermaus.junit5.internal.EnabledOnManufacturerCondition.Companion.enabled
import org.junit.jupiter.api.extension.ConditionEvaluationResult
import org.junit.jupiter.api.extension.ExecutionCondition
import org.junit.jupiter.api.extension.ExtensionContext
import org.junit.platform.commons.util.AnnotationUtils.findAnnotation
import org.junit.platform.commons.util.Preconditions

internal class DisabledOnManufacturerCondition : ExecutionCondition {

    companion object {
        private val ENABLED_BY_DEFAULT =
            ConditionEvaluationResult.enabled("@DisabledOnManufacturer is not present")
    }

    @RequiresApi(24)
    override fun evaluateExecutionCondition(context: ExtensionContext): ConditionEvaluationResult {
        val optional = findAnnotation(context.element, DisabledOnManufacturer::class.java)

        if (optional.isPresent) {
            val annotation = optional.get()
            val patterns = annotation.value
            val ignoreCase = annotation.ignoreCase

            Preconditions.condition(
                patterns.isNotEmpty(),
                "You must declare at least one Manufacturer in @DisabledOnManufacturer"
            )

            return if (patterns.any { Build.MANUFACTURER.equals(it, ignoreCase = ignoreCase) }) {
                disabled()
            } else {
                enabled()
            }
        }
        return ENABLED_BY_DEFAULT
    }
}
